NuRadioReco/modules/io/RNO_G/readRNOGDataMattak.py

```diff
@@ -0,0 +1,642 @@
+            if self._convert_to_voltage:
+               # convert adc to voltage
+               wf *= (self._adc_ref_voltage_range / (2 ** (self._adc_n_bits) - 1))
+               
+            channel.set_trace(wf, event_info.sampleRate * units.GHz)
+         
+         time_offset = get_time_offset(event_info.triggerType)
+         channel.set_trace_start_time(-time_offset)  # relative to event/trigger time
+                           
+         station.add_channel(channel)
+      
+      evt.set_station(station)
+      
+      return evt     
+
+
+   @register_run()
+   def run(self):
+      """ 
+      Loop over all events.
+      
+      Returns
+      -------
+      
+      evt: generator(NuRadioReco.framework.event)
+      """
+      event_idx = -1
+      for dataset in self._datasets:
+         dataset.setEntries((0, dataset.N()))
+                  
+         # read all event infos of the entier dataset (= run)
+         event_infos = dataset.eventInfo()
+         wfs = None
+
+         for idx, evtinfo in enumerate(event_infos):  # returns a list
+            event_idx += 1
+            
+            self.logger.debug(f"Processing event number {event_idx} out of total {self._n_events_total}")
+            t0 = time.time()
+                  
+            if self._filter_event(evtinfo, event_idx):
+               continue
+            
+            # Just read wfs if necessary
+            if wfs is None:
+               wfs = dataset.wfs()
+               
+            waveforms_of_event = wfs[idx]
+            
+            evt = self._get_event(evtinfo, waveforms_of_event)
+            
+            self._time_run += time.time() - t0
+            self.__counter += 1
+                           
+            yield evt
+
+
+
+   def get_event_by_index(self, event_index):
+      """ Allows to read a specific event identifed by its index 
+      
+      Parameters
+      ----------
+      
+      event_index: int
+         The index of a particluar event. The index is the chronological number from 0 to 
+         number of total events (across all datasets).
+         
+      Returns
+      -------
+      
+      evt: NuRadioReco.framework.event
+      """
+      
+      self.logger.debug(f"Processing event number {event_index} out of total {self._n_events_total}")
+      t0 = time.time()
+
+      dataset = self.__get_dataset_for_event(event_index)
+      event_info = dataset.eventInfo()  # returns a single eventInfo
+
+      if self._filter_event(event_info, event_index):
+         return None
+            
+      # access data
+      waveforms = dataset.wfs()
+      
+      evt = self._get_event(event_info, waveforms)
+      
+      self._time_run += time.time() - t0
+      self.__counter += 1
+   
+      return evt
+   
+      
+   def get_event(self, event_id):
+      """ Allows to read a specific event identifed by its id
+
+      Parameters
+      ----------
+
+      event_id: int
+         Event Id
+         
+      Returns
```

Comment:
Actually... :see_no_evil: can we change this to select by run number and event id? See https://github.com/nu-radio/NuRadioMC/blob/develop/NuRadioReco/modules/io/NuRadioRecoio.py#L277. If we use the reader to read multiple runs in one go, which seems to be an intended feature (?), selection by event id only is ambiguous.

Reply:
I guess you are right. Do you want to do it? Or are you fine with having that in the next PR?

Reply:
done